File: crawl.js
const { JSDOM } = require('jsdom') //JSDOM is a library that allows us to parse HTML and manipulate the DOM in Node.js. We use it to extract URLs from the HTML body.

function getURLsFromHTML(htmlBody, baseURL){
    const urls = []
    const dom = new JSDOM(htmlBody)
    //we use "a" because we want to get the href attribute of the "a" tag
    const linkElements = dom.window.document.querySelectorAll('a')
    for(const linkElement of linkElements){

        if (linkElement.href.slice(0,1) === '/') {
            //relative URL
            try{
                const urlObj = new URL(`${baseURL}${linkElement.href}`)
                urls.push(urlObj.href)
            }catch(err){
                console.log(`error with relative URL: ${err.message}`)
            }
            
        }else{
            //absolute URL

            try{
                const urlObj = new URL(linkElement.href)
                urls.push(urlObj.href)
            }catch(err){
                console.log(`error with absolute URL: ${err.message}`)
            }
            
        }
        //console.log(linkElement.href)

    }
    return urls
}

function normalizeURL(urlString) {
    const urlObj = new URL(urlString) //URL constructor alreadys lowercases hostname and pathname

    const hostPath = `${urlObj.hostname}${urlObj.pathname}`

    if (hostPath.length > 0 && hostPath.slice(-1) === '/') {
        return hostPath.slice(0, -1)
    }
    return hostPath
}

module.exports = {
    normalizeURL,
    getURLsFromHTML
}

File: crawl.py
#from bs4 import BeautifulSoup # This is a third-party library that we will use to parse HTML. We will install it using pip.
#import tldextract
#from html.parser import HTMLParser
from MyHTMLParser import MyHTMLParser
from urllib.parse import urlparse

def getURLsFromHTML(htmlBody, baseURL):
    urls = []
    dom = MyHTMLParser(htmlBody)
    linkElements = dom.findQuery("a")
    print("Link Elements: " + str(linkElements))
    for linkElement in linkElements:
        #print("Link Element: " + linkElement[:-1])
        try:
            urlObj = urlparse(baseURL + linkElement).geturl()
            print("URL Object: " + str(urlObj))
        except Exception as e:
            print("Error parsing URL: " + str(e))
            continue
        if linkElement.startswith("/"):
            
            #relative URL
            urls.append(urlObj)
        else:
            #absolute URL
            urls.append(linkElement)
    #print("Link Elements: " + str(linkElements))
    print("URLs: " + str(urls))
    return urls

def normalizeURL(url):
    urlObj = urlparse(url)
    #print("URL Object: " + str(urlObj))
    
    # .hostname naturally lowercases the hostname
    hostname = urlObj.hostname

    # strip the subdomain from the hostname
    subdomains = ["www.", "m.", "mobile.", "blog.", "news.", "shop.", "store.", "app.", "web.", "dev.", "test.", "staging."]
    for subdomain in subdomains:
        if hostname.startswith(subdomain):
            hostname = hostname[len(subdomain):]
            break

    # combine the hostname and path
    hostPath = hostname + urlObj.path
    print("Host: " + hostname)

    if len(hostPath) > 0 and hostPath.endswith('/'):
        hostPath = hostPath[:-1]


    return hostPath